nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -304,4 +329,174 @@ WaypointFollower::goalResponseCallback(
+{
+  RCLCPP_INFO(parent_node->get_logger(), "Converting GPS waypoints to Map Frame..");
+  if (!parent_node) {
+    throw std::runtime_error{
+            "Failed to lock node while trying to convert GPS waypoints to Map frame"};
+  }
+  std::vector<geometry_msgs::msg::PoseStamped> poses_in_map_frame_vector;
+  int index_of_gps_waypoints = 0;
+  for (auto && curr_gps_waypoint : gps_waypoints) {
+    auto fromLLRequest = std::make_shared<robot_localization::srv::FromLL::Request>();
+    fromLLRequest->ll_point.latitude = curr_gps_waypoint.latitude;
+    fromLLRequest->ll_point.longitude = curr_gps_waypoint.longitude;
+    fromLLRequest->ll_point.altitude = curr_gps_waypoint.altitude;
+    if (!fromll_client->wait_for_service((std::chrono::seconds(10)))) {
+      RCLCPP_ERROR(
+        parent_node->get_logger(),
+        "fromLL service from robot_localization is not available"
+        "cannot convert GPS waypoints to Map frame poses,"
+        "Make sure you have run navsat_transform_node of robot_localization");
+      return std::vector<geometry_msgs::msg::PoseStamped>();
+    }
+    RCLCPP_INFO(
+      parent_node->get_logger(), "Processing for %i'th Waypoint..",
+      index_of_gps_waypoints);
+    auto inner_client_callback =
+      [&, parent_node](rclcpp::Client<robot_localization::srv::FromLL>::SharedFuture inner_future)
+      {
+        auto result = inner_future.get();
+      };
+    auto inner_future_result = fromll_client->async_send_request(
+      fromLLRequest,
+      inner_client_callback);
+    // this poses are assumed to be on global frame (map)
+    geometry_msgs::msg::PoseStamped curr_waypoint_in_map_frame;
+    curr_waypoint_in_map_frame.header.frame_id = "map";
+    curr_waypoint_in_map_frame.header.stamp = rclcpp::Clock().now();
+    curr_waypoint_in_map_frame.pose.position.x = inner_future_result.get()->map_point.x;
+    curr_waypoint_in_map_frame.pose.position.y = inner_future_result.get()->map_point.y;
+    curr_waypoint_in_map_frame.pose.position.z = inner_future_result.get()->map_point.z;
+
+    tf2::Quaternion quat_tf;
+    quat_tf.setRPY(0, 0, 0);
+    geometry_msgs::msg::Quaternion quat_msg;
+    tf2::convert(quat_msg, quat_tf);
+    curr_waypoint_in_map_frame.pose.orientation = quat_msg;
+
+    RCLCPP_INFO(
+      parent_node->get_logger(),
+      "%i th Waypoint Long, Lat: %.8f , %.8f converted to "
+      "Map Point X,Y: %.8f , %.8f", index_of_gps_waypoints, fromLLRequest->ll_point.longitude,
+      fromLLRequest->ll_point.latitude, curr_waypoint_in_map_frame.pose.position.x,
+      curr_waypoint_in_map_frame.pose.position.y);
+    index_of_gps_waypoints++;
+    poses_in_map_frame_vector.push_back(curr_waypoint_in_map_frame);
+  }
+  RCLCPP_INFO(
+    parent_node->get_logger(),
+    "Converted all %i GPS waypoint to Map frame", poses_in_map_frame_vector.size());
+  return poses_in_map_frame_vector;
+}
+
+void WaypointFollower::GPSResultCallback(
```

Comment:
I think these 2 result callback / goal response callback functions can be shared across both action servers. I'm not 100% sure, but I think you can template this function on `ClientT` since both ClientT and ClientTGPS have the same `code` field. Then you just do `ResultCallback<T>` in the registering. So the waypoint follower class itself would be a templated class, but only these 2 functions would be using the templates. The `T` would be the type of `ClientT` with 2 implementation options: Cartesian and GPS.

Reply:
This might not be possible, but worth a look. It would certainly be cool if it did.